DTS documentation

mbolivar opened the pull request:
Re-work the introductory sections of the devicetree documentation, adding several figures and cross-references to other useful parts of the documentation.

If these changes are agreeable, I'd like to use them as a starting point to:

- more formally specify the `bindings+dts --> scripts --> C` process. This is currently mostly documented by example in dts/binding-template.yaml and the end of doc/guides/dts/index.rst.
- include API documentation for dtlib and edtlib in the docs themselves, similarly to https://docs.zephyrproject.org/latest/guides/west/west-apis.html#west-apis

My ultimate goal after that is to move on to make some changes to the way phandle+specifier-valued properties are currently managed by edtlib (and its gen_defines.py user).

mbolivar-nordic committed:
doc: add more cross-references/examples for DT overlays

Device tree overlays are a bit of a stumbling block. Try to add more
cross-references and examples for how to use them to the application
development doc and the west build page.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: add a couple of ref targets

These will be needed later.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: add more explanations, with diagrams

Re-work the introductory sections of the devicetree documentation,
adding several figures and cross-references to other useful parts of
the documentation.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
Note that the figures are included in the docs via their .PNGs to ensure the presentation is as expected. I've included the .SVGs just to make it easier for future editors. These were all generated using draw.io.

dbkinder reviewed (approved):
Very nice!

dbkinder commented:
I'm curious what issues you found that made you use .png files instead of just directly using the .svg drawings?

mbolivar commented:
Thanks for the review @dbkinder!

> I'm curious what issues you found that made you use .png files instead of just directly using the .svg drawings?

Unexpected line wrapping in lines of text inside boxes were a particular problem for me on Firefox.

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -1,177 +1,249 @@
+peripheral address in its ``reg`` property there.
+
+Application source code with a pointer to the ``struct device`` can then pass
+it to driver APIs in :zephyr_file:`zephyr/include/drivers`. These API functions
```
Nit: `zephyr/include/drivers/`, for consistency.

mbolivar replied:
>  Nit: zephyr/include/drivers/, for consistency.

Whoops, that's a bug. It should be include/drivers, not zephyr/include/drivers. I'll add the slash.

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -196,61 +269,58 @@ setting the bus speed:
+   / {
+           model = "Human readable board name";
+           compatible = "vendor,soc-on-your-board's-mcu";
+           /* rest of file */
```
(DTS supports the nicer `// rest of file` format too.)

mbolivar replied:
C89 style is what virtually all (all but three) of the .dts and .dtsi files I found in the tree are using, and Zephyr's C comment style is `/* */`, so I'm going to leave this as-is unless you're proposing a tree-wide change to C99 / C++ style single line comments.

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -196,61 +269,58 @@ setting the bus speed:
+  * Add target specific devicetree files for base SoC.  These should be
+    .dtsi files to be included in the board-specific devicetree files.
   * Add target specific YAML binding files in the dts/bindings/ directory.
     Create the yaml directory if not present.
```
Nit: Could highlight stuff like `dts/arm` and `.dtsi` above maybe.

mbolivar replied:
Sorry, I don't get it -- what do you mean by "highlight"? Emphasize (in italics, bold, etc.), wrap in `:file:`, add a `:ref:` to somewhere earlier in the doc?

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -284,7 +354,7 @@ The following is a more precise list of required files:
 
   * Overlay files contain tweaks or changes to the SoC and Board support files
-    described above. They can be used to modify Device Tree configurations
+    described above. They can be used to modify Devicetree configurations
```
Maybe `Devicetree` could be consistently lowercase too, except for titles (e.g. `Devicetree specification`) and stuff. Less thinking required.

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -1,177 +1,249 @@
 
-System build requirements
-*************************
+   /ancestor-1-name/ancestor-2-name/node-name
```
Maybe a short generic example could be used here, like

```
/dts-v1/;

/ {
	foo = "string";
	node {
		bar = <3>;
	};
};
```

The explanations could then refer back to it when explaining paths and properties.

mbolivar replied:
Good idea. I'll rework this.

ulfalizer commented on doc/guides/dts/index.rst:
```diff
@@ -1,177 +1,249 @@
+devices.  These fixup files by default reside in the board and soc directories
+and are named ``dts_fixup.h``.  These fixup files map the generated include
+information to the current driver/source usage. They exist for historical
+reasons; Zephyr is moving away from needing or using these files.
```
Nit: Single space after period for consistency.

ulfalizer reviewed (approved):
Great to have more documentation for this.

Some random nits from skimming. Not super important, and can be tweaked later.

pabigot commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -151,37 +161,44 @@ To let west decide for you if a pristine build is needed, use ``-p auto``::
-As a final example, to merge the :file:`file.conf` Kconfig fragment into your
-build's :file:`.config`::
+To use the file :file:`enable-modem.overlay` as a :ref:`device tree overlay
+<device-tree>` by setting :ref:`DTC_OVERLAY_FILE <application_dt>`::
```
s/by setting/set/

mbolivar replied:
I think the grammar is correct as is (to accomplish X by setting Y to Z), but it must be unclear. How about this?

```
To set :ref:`DTC_OVERLAY_FILE <application_dt>` to :file:`enable-modem.overlay`,
using that file as a :ref:`device tree overlay <device-tree>`::
```

pabigot reviewed (approved):
Thanks for doing this.

carlescufi reviewed (approved):
This looks great @mbolivar. Thanks for the PR.
I subscribe to several of the comments that @ulfalizer posted, but otherwise this is already a massive improvement.
A preview can be found here:
https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/19630/docs/guides/dts/index.html

carlescufi commented:
Also, @mbolivar I think it would be good to add a bit more info about the following:

-  `generated_dts_board.h`, which is referred to in several places, does not really exist. Instead we have `./include/generated/generated_dts_board_fixups.h` and `./include/generated/generated_dts_board_unfixed.h`. This is very confusing to me and I feel we are lacking clarification.

- The contents of `generated_dts_board.h` are not mentioned either. We should explain what kind of macros users should expect to be generated there, since this would help those users puzzled to see a build failure mention `DT_INST_<>` for example.

mbolivar commented on doc/application/index.rst:
```diff
@@ -1492,15 +1493,17 @@ docstrings at the top of ``scripts/kconfig/menuconfig.py`` and
+In some cases the information contained in Device Tree files is closely
+connected to the software and might need to be modified using the overlay file
+concept. This can be relevant for many of the different Device Tree nodes, but
+is particularly useful for :ref:`certain types of nodes <dt-alias-chosen>`.
```
> Missed "device tree" -> "devicetree" stuff in this file

Intentionally omitted to keep the PR scope smaller. I'll add a patch on top that changes this tree-wide.

mbolivar commented on doc/guides/dts/index.rst:
```diff
@@ -1,177 +1,249 @@
+with any overlays into a final :file:`BOARD.dts_compiled` file in the
+application build directory.  :file:`BOARD.dts_compiled` is the final
+devicetree source. It is also in DTS format. More information on ``dtc`` can be
+found in the `dtc source code repository`_.
```
> Maybe this could mention that the files are just concatenated (it's done here), and that multiple definitions for a node get merged in DTS.

This was also intentionally left out to keep PR scope smaller. This falls into the "more formally specify the bindings+dts --> scripts --> C process" follow-up work mentioned in the cover letter.

> The new script runs directly on the .dts files btw, and don't need dtc. The old scripts are run on the output of dtc -Odts <.dts file>, because they don't deal with things like replacing phandle (... = <&foo>;) and path (... = &foo;) references.

OK, I'll update the diagram, thanks.

> There's also a native DTS syntax for including files btw: /include/ foo. It doesn't seem to get used that often though.

Yep, I know. I didn't mention it precisely because it doesn't get used by Zephyr.

mbolivar commented on doc/guides/dts/index.rst:
```diff
@@ -1,177 +1,249 @@
+include the generated C headers in the build directory directly -- use
+``generated_dts_board.h`` instead. A file named
+:file:`generated_dts_board.conf` is also placed in the build directory; it is
+mostly useful for inspecting the results of Zephyr's DTS scripts.
```
> We're trying to get rid of it though.

OK, if this file is on its way out, I'll just remove the mention of it here, thanks.

mbolivar replied:
By the way, information like this ("we're trying to get rid of [generated_dts_board.conf]") is exactly the sort of thing I'd like to know as a developer who uses DT, so this exercise is already bearing fruit :+1:

mbolivar-nordic committed:
doc: s/device tree/devicetree/

DTSpec writes this as a single word, presumably to make it easier to
grep for / more precise. Follow along in the rest of the docs now that
our main DT docs page agrees with this usage.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> * `generated_dts_board.h`, which is referred to in several places, does not really exist as a build product. Instead we have `./include/generated/generated_dts_board_fixups.h` and `./include/generated/generated_dts_board_unfixed.h`. This is very confusing to me and I feel we are lacking clarification.
>

@carlescufi I've tried to improve this in "input and output files".

> * The contents of `generated_dts_board.h` are not mentioned either. We should explain what kind of macros users should expect to be generated there, since this would help those users puzzled to see a build failure mention `DT_INST_<>` for example.

This is follow-up work under the "more formally specify the bindings+dts --> scripts --> C process" umbrella, so I'd like to defer it until after this PR.

mbolivar commented:
@ulfalizer comments addressed, thanks again for review.

mbolivar-nordic committed:
doc: dts: move example to the right place

The K6X example is part of the wrong subsection (devicetree vs
kconfig) currently. Move it up to the right place (the section which
was recently renamed to "input and output files")

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>
